src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -131,33 +122,43 @@ where s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id)
 
         ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
 
-        int version;
-
         class VersionedSagaEntity
         {
             public VersionedSagaEntity(IContainSagaData sagaData, VersionedSagaEntity original = null)
             {
                 SagaData = DeepClone(sagaData);
-                versionCache = original?.versionCache ?? new ConditionalWeakTable<IContainSagaData, SagaVersion>();
+                if (original != null)
+                {
+                    original.ConcurrencyCheck(sagaData);
+
+                    versionCache = original.versionCache;
+                    version = original.version;
+                    version++;
```

Comment:
is there an issue if you end up creating 2 x VersionedSagaEntity based on the same `original` since both new instances will end up with the same version (and also share the same `versionCache`)?